### PR TITLE
Add initial version of object counting metrics.

### PR DESCRIPTION
**Pull request by lolepezy**
This is to add basic Prometheus metrics to the number of objects that are published or withdrawn.

### CODE REVIEW

File: src/main/scala/net/ripe/rpki/publicationserver/metrics/Metrics.scala [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+  val countPublishedObjects = Counter
+    .build()
+    .name("objects_published")
+    .help("Number of objects published by publishing clients")
+    .register(registry)
+
+  val countWithdrawnObjects = Counter
+    .build()
+    .name("objects_withdrawn")
```
Comment (ties) [inline src/main/scala/net/ripe/rpki/publicationserver/metrics/Metrics.scala:31]:
```suggestion
    .name("rpkipublicationserver_objects_withdrawn_total")
```

The prefix and `total` (+unit where applicable) applies to all of the metrics, [it is recommended that](https://prometheus.io/docs/practices/naming/):
  * metrics are prefixed with the application name, I would use `rpkipublicationserver` here
  * metrics are followed by a unit if applicable, counters are followed by `_total`

File: src/main/scala/net/ripe/rpki/publicationserver/StateActor.scala [hunk 2]
```diff
@@ -104,16 +110,20 @@ class StateActor(conf: AppConfig) extends Actor with Hashing with Logging {
...
+          val newHash = hash(newBytes)
           logger.debug(s"Replacing $uri with hash $foundHash -> $newHash")
-          Right(state + (uri -> (bytes, newHash, clientId)))
+          // one added one deleted
+          metrics.publishedObject()
+          metrics.withdrawnObject()
+          Right(state + (uri -> (newBytes, newHash, clientId)))
         } else
           Left(ReportError(BaseError.NonMatchingHash,
             Some(s"Cannot republish the object [$uri], hash doesn't match, passed ${hashToReplace.toUpperCase}, but existing one is ${foundHash.toUpperCase}.")))
```
Comment (ties) [inline src/main/scala/net/ripe/rpki/publicationserver/StateActor.scala:129]: Do we want to count these failures?

File: src/main/scala/net/ripe/rpki/publicationserver/StateActor.scala [hunk 3]
```diff
@@ -125,10 +135,10 @@ class StateActor(conf: AppConfig) extends Actor with Hashing with Logging {
   def applyCreate(state: State, clientId: ClientId, uri: URI, bytes: Bytes): Either[ReportError, State] = {
     if (state.contains(uri)) {
       Left(ReportError(BaseError.HashForInsert, Some(s"Tried to insert existing object [$uri].")))
```
Comment (ties) [inline src/main/scala/net/ripe/rpki/publicationserver/StateActor.scala:137]: As well as these?

File: src/main/scala/net/ripe/rpki/publicationserver/metrics/Metrics.scala [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import net.ripe.rpki.publicationserver.store.ObjectStore.State
+import java.io.StringWriter
+import io.prometheus.client.exporter.common.TextFormat
+import io.prometheus.client._
+
+class Metrics(val registry: CollectorRegistry) {
+
+  val countPublishedObjects = Counter
+    .build()
+    .name("objects_published")
```
Comment (ties) [inline src/main/scala/net/ripe/rpki/publicationserver/metrics/Metrics.scala:25]: `published`, `withdrawn` could be tags on one metric?

That makes querying easier.

File: src/main/scala/net/ripe/rpki/publicationserver/metrics/Metrics.scala [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+  val countWithdrawnObjects = Counter
+    .build()
+    .name("objects_withdrawn")
+    .help("Number of objects published by publishing clients")
+    .register(registry)
+
+  val lastTimeReceived = Gauge
+    .build()
+    .name("last_object_recived")
```
[the remark this answers is not in the document]
Comment (lolepezy) [reply src/main/scala/net/ripe/rpki/publicationserver/metrics/Metrics.scala:37]: rpkipublicationserver_last_object_received?
No sure what's the use of splitting in this case, probably doesn't give much new information.

File: src/main/scala/net/ripe/rpki/publicationserver/StateActor.scala (hunk 2 above)
Comment (lolepezy) [reply src/main/scala/net/ripe/rpki/publicationserver/StateActor.scala:129]: Yes, I think so. Will add counter for then as well.

File: src/main/scala/net/ripe/rpki/publicationserver/StateActor.scala (hunk 3 above)
Comment (lolepezy) [reply src/main/scala/net/ripe/rpki/publicationserver/StateActor.scala]: Yeap, done.

File: src/main/scala/net/ripe/rpki/publicationserver/metrics/Metrics.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (lolepezy) [reply src/main/scala/net/ripe/rpki/publicationserver/metrics/Metrics.scala]: Implemented in a split version: add, replace, delete failures separately.

File: src/main/scala/net/ripe/rpki/publicationserver/metrics/Metrics.scala (hunk 4 above)
Comment (lolepezy) [reply src/main/scala/net/ripe/rpki/publicationserver/metrics/Metrics.scala]: Will do, need to look through the prometheus documentation.

File: build.sbt [hunk 6]
```diff
@@ -103,7 +103,7 @@ publishTo := {
   if (sys.props.isDefinedAt("build.number"))
     Some(Resolver.file("",  new File(Path.userHome.absolutePath+"/.m2/repository")))
   else
-    Some("ripe-snapshots" at "http://nexus.ripe.net/nexus/content/repositories/snapshots")
+    Some("ripe-snapshots" at "https://nexus.ripe.net/nexus/content/repositories/snapshots")
```
Comment (wibisono) [inline build.sbt:106]: While we're at this line 40 also generates warning and needs https: `resolvers += "Codehaus Maven2 Repository" at "http://repository.codehaus.org/"`

Comment (lolepezy) [reply build.sbt:106]: Yeap, fixed.

Comment (ties) [reply build.sbt:106]: Nice when it's quieter :d

File: src/test/scala/net/ripe/rpki/publicationserver/integration/PublicationIntegrationTest.scala [hunk 7]
```diff
@@ -146,21 +144,25 @@ class PublicationIntegrationTest
...
-    metrics3 should include("rpkipublicationserver_objects_failedtoreplace_total 1.0")
-    metrics3 should include("rpkipublicationserver_objects_failedtowithdraw_total 2.0")
+    forMetrics { metrics => 
+        metrics should include("""rpkipublicationserver_object_operations_total{operation="publish",} 4.0""")
+        metrics should include("""rpkipublicationserver_object_operations_total{operation="withdraw",} 2.0""")        
+    }
     
   }  
 
+  def forMetrics(f: String => Unit) = {
```
Comment (wibisono) [inline src/test/scala/net/ripe/rpki/publicationserver/integration/PublicationIntegrationTest.scala:162]: neat

**Pull request merged**